src/h3/_cy/cells.pyx

```diff
@@ -0,0 +1,281 @@
+cimport h3lib
+from .h3lib cimport bool, int64_t, H3int
+from libc cimport stdlib
+
+from .util cimport (
+    check_cell,
+    check_res,
+    create_ptr,
+    create_mv,
+    empty_memory_view, # want to drop this import if possible
+)
+
+from .util import H3ValueError
+
+# todo: add notes about Cython exception handling
+
+
+# bool is a python type, so we don't need the except clause
+cpdef bool is_cell(H3int h):
+    """Validates an H3 cell (hexagon or pentagon)
+
+    Returns
+    -------
+    boolean
+    """
+    return h3lib.h3IsValid(h) == 1
+
+cpdef bool is_pentagon(H3int h):
+    return h3lib.h3IsPentagon(h) == 1
+
+cpdef int get_base_cell(H3int h) except -1:
+    check_cell(h)
+
+    return h3lib.h3GetBaseCell(h)
+
+
+cpdef int resolution(H3int h) except -1:
+    """Returns the resolution of an H3 Index
+    0--15
+    """
+    check_cell(h)
+
+    return h3lib.h3GetResolution(h)
+
+
+cpdef H3int parent(H3int h, res=None) except 1:
+    check_cell(h)
+
+    if res is None:
+        res = resolution(h) - 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    check_res(res)
+
+    return h3lib.h3ToParent(h, res)
+
+
+cpdef int distance(H3int h1, H3int h2) except -1:
+    """ compute the hex-distance between two hexagons
+    """
+    check_cell(h1)
+    check_cell(h2)
+
+    d = h3lib.h3Distance(h1,h2)
+
+    return d
+
+cpdef H3int[:] disk(H3int h, int k):
+    """ Return cells at grid distance `<= k` from `h`.
+    """
+    check_cell(h)
+    if k < 0:
+        raise H3ValueError('Invalid ring size: {}'.format(k))
+
+    n = h3lib.maxKringSize(k)
+
+    ptr = create_ptr(n) # todo: return a "smart" pointer that knows its length?
+    h3lib.kRing(h, k, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+
+cpdef H3int[:] _ring_fallback(H3int h, int k):
+    """
+    `ring` tries to call `h3lib.hexRing` first; if that fails, we call
+    this function, which relies on `h3lib.kRingDistances`.
+
+    Failures for `h3lib.hexRing` happen when that alg runs into a pentagon.
+    """
+    check_cell(h)
+
+    n = h3lib.maxKringSize(k)
+    # array of h3 cells
+    ptr = create_ptr(n)
+
+    # array of cell distances from `h`
+    dist_ptr = <int*> stdlib.calloc(n, sizeof(int))
+    if dist_ptr is NULL:
+        raise MemoryError()
+
+    h3lib.kRingDistances(h, k, ptr, dist_ptr)
+
+    distances = <int[:n]> dist_ptr
+    distances.callback_free_data = stdlib.free
+
+    for i,v in enumerate(distances):
+        if v != k:
+            ptr[i] = 0
+
+    mv = create_mv(ptr, n)
+
+    return mv
+
+cpdef H3int[:] ring(H3int h, int k):
+    """ Return cells at grid distance `== k` from `h`.
+    Collection is "hollow" for k >= 1.
+    """
+    check_cell(h)
+
+    n = 6*k if k > 0 else 1
```

Comment:
do we need a check for k < 0?

Reply:
nice catch! I added 2 tests, a new error type, and some checks